rust/xaynet-client/src/state_machine/phases/sum.rs

```diff
@@ -0,0 +1,77 @@
+use xaynet_core::{
+    crypto::{EncryptKeyPair, Signature},
+    message::Sum as SumMessage,
+};
+
+use super::{Phase, Progress, State, Step, Sum2};
+use crate::{
+    state_machine::{io::StateMachineIO, TransitionOutcome},
+    MessageEncoder,
+};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Sum {
+    pub ephm_keys: EncryptKeyPair,
+    pub sum_signature: Signature,
+    pub message: Option<MessageEncoder>,
+}
+
+impl Sum {
+    pub fn new(sum_signature: Signature) -> Self {
+        Sum {
+            ephm_keys: EncryptKeyPair::generate(),
+            sum_signature,
+            message: None,
+        }
+    }
+}
+
+#[async_trait]
+impl<IO> Step<IO> for Phase<Sum, IO>
+where
+    IO: StateMachineIO,
+{
+    async fn step(mut self) -> TransitionOutcome<IO> {
+        info!("sum task");
+
+        self = try_progress!(self.compose_sum_message());
+
+        // FIXME: currently if sending fails, we lose the message,
+        // thus wasting all the work we've done in this phase
+        let message = self.state.phase.message.take().unwrap();
```

Comment:
Does it make sense to clone the message?

Reply:
I think it would be a little wasteful. I'd like to come up with a more general solution for messages. I have ideas, but it would be too much for that PR.